File: contracts/api/clients/jakarta/src/test/java/app/epistola/client/jakarta/error/ProblemDetailParserTest.java
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

package app.epistola.client.jakarta.error;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import app.epistola.client.jakarta.model.DataModelValidationError;
import app.epistola.client.jakarta.model.ValidationError;
import jakarta.ws.rs.core.MediaType;
import java.util.List;
import org.junit.jupiter.api.Test;

class ProblemDetailParserTest {

    @Test
    void reads_the_base_problem_members() {
        var parsed = ProblemDetailParser.parse(
                """
                {
                  "type": "https://epistola.app/errors/not-found",
                  "title": "Not Found",
                  "status": 404,
                  "detail": "Template 'monthly-invoice' does not exist",
                  "instance": "/api/tenants/acme-corp/catalogs/default/templates/monthly-invoice"
                }
                """);

        assertNotNull(parsed);
        assertEquals("https://epistola.app/errors/not-found", parsed.problem().getType().toString());
        assertEquals("Not Found", parsed.problem().getTitle());
        assertEquals(404, parsed.problem().getStatus());
        assertEquals("Template 'monthly-invoice' does not exist", parsed.problem().getDetail());
        assertTrue(parsed.errors().isEmpty());
        assertTrue(parsed.validationErrors().isEmpty());
    }

    @Test
    void reads_the_errors_extension_of_a_validation_problem() {
        var parsed = ProblemDetailParser.parse(
                """
                {
                  "type": "https://epistola.app/errors/validation-error",
                  "title": "Validation Failed",
                  "status": 400,
                  "errors": [
                    {"field": "id", "message": "must match pattern", "rejectedValue": "ACME"},
                    {"field": "name", "message": "must not be blank"}
                  ]
                }
                """);

        assertNotNull(parsed);
        List<ValidationError> errors = parsed.errors();
        assertEquals(2, errors.size());
        assertEquals("id", errors.get(0).getField());
        assertEquals("must match pattern", errors.get(0).getMessage());
        assertEquals("ACME", errors.get(0).getRejectedValue());
        assertEquals("name", errors.get(1).getField());
        assertTrue(parsed.validationErrors().isEmpty());
    }

    @Test
    void reads_the_validation_errors_extension_of_a_data_model_problem() {
        var parsed = ProblemDetailParser.parse(
                """
                {
                  "type": "https://epistola.app/errors/data-model-validation-error",
                  "title": "Data Model Validation Failed",
                  "status": 422,
                  "validationErrors": {
                    "happy-path": [
                      {"path": "/customer/name", "message": "is required"},
                      {"path": "/total", "message": "must be a number"}
                    ],
                    "empty-invoice": [
                      {"path": "/lineItems", "message": "must have at least 1 item"}
                    ]
                  }
                }
                """);

        assertNotNull(parsed);
        assertEquals(2, parsed.validationErrors().size());
        List<DataModelValidationError> happyPath = parsed.validationErrors().get("happy-path");
        assertEquals(2, happyPath.size());
        assertEquals("/customer/name", happyPath.get(0).getPath());
        assertEquals("is required", happyPath.get(0).getMessage());
        assertEquals(1, parsed.validationErrors().get("empty-invoice").size());
        assertTrue(parsed.errors().isEmpty());
    }

    @Test
    void ignores_extension_members_the_contract_does_not_model() {
        var parsed = ProblemDetailParser.parse(
                """
                {
                  "type": "https://epistola.app/errors/rate-limited",
                  "title": "Too Many Requests",
                  "status": 429,
                  "retryAfterSeconds": 30,
                  "quota": {"limit": 100, "window": "1m"}
                }
                """);

        assertNotNull(parsed);
        assertEquals("Too Many Requests", parsed.problem().getTitle());
    }

    @Test
    void defaults_the_type_to_about_blank_when_absent() {
        var parsed = ProblemDetailParser.parse("{\"title\":\"Internal Server Error\",\"status\":500}");

        assertNotNull(parsed);
        assertEquals(ProblemTypes.BLANK_TYPE, parsed.problem().getType());
        assertNull(ProblemTypes.slugFor(parsed.problem().getType()));
    }

    @Test
    void returns_null_for_bodies_that_are_not_usable_problem_documents() {
        assertNull(ProblemDetailParser.parse(null));
        assertNull(ProblemDetailParser.parse(""));
        assertNull(ProblemDetailParser.parse("   "));
        assertNull(ProblemDetailParser.parse("not json at all"));
        assertNull(ProblemDetailParser.parse("{\"unterminated\": "));
        assertNull(ProblemDetailParser.parse("[{\"title\":\"an array is not a problem\"}]"));
    }

    @Test
    void recognises_the_problem_media_type_and_only_that() {
        assertTrue(ProblemDetailParser.isProblemJson(MediaType.valueOf("application/problem+json")));
        assertTrue(ProblemDetailParser.isProblemJson(MediaType.valueOf("application/problem+json;charset=UTF-8")));
        assertTrue(ProblemDetailParser.isProblemJson(MediaType.valueOf("APPLICATION/PROBLEM+JSON")));

        assertFalse(ProblemDetailParser.isProblemJson(null));
        assertFalse(ProblemDetailParser.isProblemJson(MediaType.APPLICATION_JSON_TYPE));
        assertFalse(ProblemDetailParser.isProblemJson(MediaType.valueOf("application/vnd.epistola.v1+json")));
        assertFalse(ProblemDetailParser.isProblemJson(MediaType.TEXT_PLAIN_TYPE));
    }
}

File: contracts/api/clients/nodejs-fetch/test/error/problemDetailParser.test.ts
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

import assert from 'node:assert/strict'
import { test } from 'node:test'
import { KnownProblemSlugs, ProblemDetailException, ResponseError, isProblemJson, parseProblem } from '../../src/index.js'

function parse(body: unknown, status: number): ProblemDetailException | undefined {
  const text = typeof body === 'string' ? body : JSON.stringify(body)
  return parseProblem(text, status, new Response(text, { status, headers: { 'Content-Type': 'application/problem+json' } }))
}

test('parses a base problem and exposes the type slug', () => {
  const problem = parse({ type: 'https://epistola.app/errors/not-found', title: 'Not Found', status: 404, detail: 'tenant acme not found', instance: '/api/tenants/acme' }, 404)!
  assert.ok(problem instanceof ProblemDetailException)
  assert.ok(problem instanceof ResponseError)
  assert.equal(problem.typeSlug, KnownProblemSlugs.NOT_FOUND)
  assert.equal(problem.type, 'https://epistola.app/errors/not-found')
  assert.equal(problem.title, 'Not Found')
  assert.equal(problem.problemStatus, 404)
  assert.equal(problem.statusCode, 404)
  assert.equal(problem.detail, 'tenant acme not found')
  assert.equal(problem.instance, '/api/tenants/acme')
  assert.equal(problem.message, '404 Not Found: tenant acme not found')
  assert.equal(problem.isValidationProblem, false)
  assert.equal(problem.isDataModelValidationProblem, false)
  assert.deepEqual(problem.extensions, {})
})

test('parses a validation problem with its errors array', () => {
  const problem = parse({
    type: 'https://epistola.app/errors/validation-error',
    title: 'Validation Failed',
    status: 400,
    errors: [
      { field: 'name', message: 'must not be blank' },
      { field: 'slug', message: 'invalid', rejectedValue: 'BAD' },
    ],
  }, 400)!
  assert.equal(problem.typeSlug, KnownProblemSlugs.VALIDATION_ERROR)
  assert.ok(problem.isValidationProblem)
  assert.deepEqual(problem.errors.map((e) => e.field), ['name', 'slug'])
  assert.equal(problem.errors[1].rejectedValue, 'BAD')
  // The raw extension member is also reachable by name.
  assert.ok(Array.isArray(problem.extensions.errors))
})

test('parses a data-model validation problem with its validationErrors map', () => {
  const problem = parse({
    type: 'https://epistola.app/errors/data-model-validation-error',
    title: 'Unprocessable',
    status: 422,
    validationErrors: { 'example-a': [{ path: '#/customer/name', message: 'required' }] },
  }, 422)!
  assert.equal(problem.typeSlug, KnownProblemSlugs.DATA_MODEL_VALIDATION_ERROR)
  assert.ok(problem.isDataModelValidationProblem)
  assert.equal(problem.validationErrors['example-a'][0].path, '#/customer/name')
})

test('keeps extension members the contract does not name', () => {
  const problem = parse({ type: 'https://epistola.app/errors/catalog-schema-too-old', title: 'Too old', status: 409, version: 4, baselineVersion: 6 }, 409)!
  assert.equal(problem.typeSlug, 'catalog-schema-too-old')
  assert.deepEqual(problem.extensions, { version: 4, baselineVersion: 6 })
})

test('about:blank and a missing type have no slug', () => {
  const blank = parse({ type: 'about:blank', title: 'Server Error', status: 500 }, 500)!
  assert.equal(blank.typeSlug, undefined)
  assert.equal(blank.type, 'about:blank')
  const missing = parse({ title: 'Server Error', status: 500 }, 500)!
  assert.equal(missing.type, 'about:blank')
  assert.equal(missing.typeSlug, undefined)
})

test('a body missing title or status falls back to the response status', () => {
  const problem = parse({ type: 'https://epistola.app/errors/rate-limited' }, 429)!
  assert.equal(problem.problemStatus, 429)
  assert.equal(problem.message, '429 429')
})

test('malformed JSON and non-object bodies yield nothing', () => {
  assert.equal(parse('{ not json', 400), undefined)
  assert.equal(parse('[1, 2, 3]', 400), undefined)
  assert.equal(parse('null', 400), undefined)
})

test('isProblemJson ignores parameters and case', () => {
  assert.ok(isProblemJson('application/problem+json'))
  assert.ok(isProblemJson('Application/Problem+JSON; charset=utf-8'))
  assert.ok(!isProblemJson('application/json'))
  assert.ok(!isProblemJson(undefined))
})
